Add parsing of MSBuild SDK dependencies (NuGet)

Zastai opened the pull request:
This adds the three ways to specify an MSBuild SDK reference inside a project file, including test cases for each (parsing and updating).

Intended to fix #2839 

Note: I'm new to Ruby, but do have access to the JetBrainz IDEs, which has helped. So please forgive novice code issues.

Zastai commented:
Hmm - looks like I missed that while the actual tests pass, the coverage is complaining about a 0.02% drop.
Will see what I can do about that.

Zastai commented:
Not a lot, it seems. Well, I could remove two existing-but-unused methods (`declaration_nodes` in the two finders), but that has nothing to do with what this PR is about, so seems like the wrong way to go.

Zastai commented:
Would be nice to get some feedback on this...

jurre commented:
Thanks for doing this work @Zastai, I'll try to make some time next week to thoroughly review this and get a bit more familiar with MSBuild SDK references. Just wanted to let you know that we appreciate the contribution.

japj commented:
Hello, great to see this (I recently “discovered” dependabot).
We have been building our own MSBuild Sdk for a while now and we actually use a different way to specify the sdk version (that is documented in the link you provided https://docs.microsoft.com/en-us/visualstudio/msbuild/how-to-use-project-sdk)

The global.json can have an msbuild-sdks section that can be modified to “upgrade” the sdk for all csproj files that use that sdk:

```
{
    "msbuild-sdks": {
        "My.Custom.Sdk": "5.0.0",
        "My.Other.Sdk": "1.0.0-beta"
    }
}

```

so in that case, just modifying the global.json is enough.
Could you perhaps add that behavior to this PR?

japj commented:
Sorry, I just read #2839 where you mention that global.json support is already in?

Zastai commented:
That was the plan - but then I found that global.json already had that behaviour: https://github.com/dependabot/dependabot-core/blob/3bca853184b6a345fbee4a4cf2e9d57072ebe90b/nuget/lib/dependabot/nuget/file_parser/global_json_parser.rb#L23

Zastai commented:
@jurre Any update?

Zastai commented:
@jurre
It has now been 8 months since I created this PR. Would be nice to get this properly reviewed.

Zastai commented:
We're rapidly approaching this PR's one year anniversary...
Any chance of a review, @jurre?

jurre commented on nuget/lib/dependabot/nuget/file_updater/project_file_declaration_finder.rb:
```diff
@@ -95,6 +111,80 @@ def declaring_file
+          dep_name = dependency_name&.downcase
+          dep_version = declaring_requirement.fetch(:requirement)
+          strings = []
+          declaring_file.content.scan(SDK_PROJECT_REGEX).each do |string|
+            xml = string
```
```suggestion
          declaring_file.content.scan(SDK_PROJECT_REGEX).each do |xml|
```

Zastai replied:
Applied

jurre commented on nuget/lib/dependabot/nuget/file_updater/project_file_declaration_finder.rb:
```diff
@@ -95,6 +111,80 @@ def declaring_file
+          dep_name = dependency_name&.downcase
+          dep_version = declaring_requirement.fetch(:requirement)
+          strings = []
+          declaring_file.content.scan(regex).each do |string|
+            xml = string
```
```suggestion
          declaring_file.content.scan(regex).each do |xml|
```

Zastai replied:
Applied

Zastai replied:
Actually, I'm having to revert this - `string` needs to be used unchanged afterwards, while `xml` potentially gets modified.

jurre replied:
fwiw string will also get modified:

```ruby
irb(main):004:0> string = "foo"
irb(main):005:0> xml = string
irb(main):006:0> xml << "x"
=> "foox"
irb(main):007:0> string
=> "foox"
```

Zastai replied:
It would if I used `<<`, but I use `xml +=` and it's my understanding that that creates a new string.

jurre replied:
Ah that's correct, but looking at the code we never refer to `string` after `xml` is changed, at least not now that the code is extracted to `parse_element`. Anyway, it's not really important I guess

Zastai replied:
We do in `sdk_strings`: `strings << string`; that's why parse_element() retains the `+=`.

jurre commented on nuget/lib/dependabot/nuget/file_updater/project_file_declaration_finder.rb:
```diff
@@ -95,6 +111,80 @@ def declaring_file
+          strings = []
+          declaring_file.content.scan(SDK_PROJECT_REGEX).each do |string|
+            xml = string
+            xml += "</Project>" unless string.end_with?("/>")
+            node = Nokogiri::XML(xml)
+            node.remove_namespaces!
+            element = node.at_xpath("/Project")
+            next unless element
```
There's some duplication between this and [the block in `#sdk_strings`](https://github.com/dependabot/dependabot-core/pull/2849/files#diff-fa02b4ba56ca051e40b58c9152d7796a54f9f2f017a304fc911a7edb41aee838R164-R168), could it be worth extracting a helper method?

Zastai replied:
Will have a look

Zastai replied:
This has been refactored, with several ignore-complexity comments removed.

jurre commented on nuget/lib/dependabot/nuget/file_parser/project_file_parser.rb:
```diff
@@ -50,6 +55,63 @@ def dependency_set(project_file:)
+            name = import_node.attribute("Sdk")&.value&.strip
+            version = import_node.attribute("Version")&.value&.strip
+
+            dependency =
+              build_dependency(name, version, version, nil, project_file)
```
```suggestion
            dependency = build_dependency(name, version, version, nil, project_file)
```

We relaxed some linter rules a while ago. This is not important but I find it slightly easier on the eyes, feel free to ignore though.

Zastai replied:
Applied (to all 3 such splits)

jurre commented on nuget/lib/dependabot/nuget/file_parser/project_file_parser.rb:
```diff
@@ -50,6 +55,63 @@ def dependency_set(project_file:)
+            name = sdk_node.attribute("Name")&.value&.strip
+            version = sdk_node.attribute("Version")&.value&.strip
+
+            dependency =
+              build_dependency(name, version, version, nil, project_file)
```
```suggestion
            dependency = build_dependency(name, version, version, nil, project_file)
```

We relaxed some linter rules a while ago. This is not important but I find it slightly easier on the eyes, feel free to ignore though.

jurre commented on nuget/lib/dependabot/nuget/file_updater/project_file_declaration_finder.rb:
```diff
@@ -95,6 +111,80 @@ def declaring_file
+              strings << sdk_reference
+            end
+          end
+          # Drop duplicates, to avoid potential issues with repeated replacements (e.g. upgrading Foo/1.0.0-beta to
+          # Foo/1.0.0-beta.2). However, that turns the empty array into nil.
+          strings.uniq! || []
```
```suggestion
          strings.uniq
```

The `uniq!` version of the method mutates the array in-place, but since we're returning the value it's fine to return a new instance, and this so called "non bang" version of the method does not return an empty array into nil.

Zastai replied:
Applied

jurre commented on nuget/lib/dependabot/nuget/file_updater/project_file_declaration_finder.rb:
```diff
@@ -95,6 +111,80 @@ def declaring_file
+          sdk_strings(SDK_SDK_REGEX, "Sdk", "Name", "Version")
+        end
+
+        # rubocop:disable Metrics/CyclomaticComplexity
+        # rubocop:disable Metrics/PerceivedComplexity
```
I think there might be some room to extract certain parts of this into separate methods, the linter things the complexity of this method is too high and I kind of agree that it's a little hard to follow.

I do also understand that you've submitted this PR a really long time ago and you might not have an appetite to dig back in and refactor things, but it would be appreciated if we can clean this up a little and reduce the complexity of the method a little.

Zastai replied:
Done

jurre reviewed (commented):
Overall these changes look good, I have made some mostly superficial suggestions regarding readability and rubyisms.

I am not super familiar with how the newly added functionality is typically used in NuGet so I'm having somewhat of a hard time reviewing for correctness.

I think @xlgmokha has some more experience on this so I would love to get his eyes on this as well.

Do you happen to have a sample project I could test this end-to-end on?

Apologies for the slow response on this on my part, it definitely fell through the cracks a couple of times and going to try to land this ASAP.

Zastai commented:
I should be able to make a recipe for setting up an otherwise-empty project with each form of reference, if that suits.
All it would need is an installed .NET SDK (or .NET Core SDK).

Zastai commented:
I think I addressed all your remarks, @jurre.

Sample projects are present as fixtures, but in order to set up a repo for "real" testing, here are the recipes.

## Recipes

Each recipe starts from an empty directory and sets up a dependency on an old version of an SDK package I maintain (`MetaBrainz.Build.Sdk`).
Dependabot should suggest an update for it.

### A. Using global.json (already supported by dependabot)
1. Create a `.csproj` file (name can be chosen freely) containing
   ```
   <Project Sdk="MetaBrainz.Build.Sdk">
   </Project>
   ```
2. Create a `global.json` file containing
   ```
   {
     "msbuild-sdks": {
       "MetaBrainz.Build.Sdk" : "1.0.0"
     }
   }
   ```

### B. Using a version specifier in the Project tag (1)
1. Create a `.csproj` file (name can be chosen freely) containing
   ```
   <Project Sdk="MetaBrainz.Build.Sdk/1.0.0">
   </Project>
   ```

### C. Using a version specifier in the Project tag (2)
This is the same case as B, but it checks for support for multiple SDK specifications.
1. Create a `.csproj` file (name can be chosen freely) containing
   ```
   <Project Sdk="Microsoft.NET.Sdk;MetaBrainz.Build.Sdk/1.0.0">
   </Project>
   ```

### D. Using an Sdk tag
1. Create a `.csproj` file (name can be chosen freely) containing
   ```
   <Project>

     <Sdk Name="MetaBrainz.Build.Sdk" Version="1.0.0" />

   </Project>
   ```

### E. Using an Sdk attribute on an Import tag
1. Create a `.csproj` file (name can be chosen freely) containing
   ```
   <Project>

     <Import Sdk="MetaBrainz.Build.Sdk" Project="Defaults.props" Version="1.0.0" />
     <Import Project="CSharp.props" Version="1.0.0" Sdk="MetaBrainz.Build.Sdk" />

     <Target Name="Dummy" />

   </Project>
   ```

Zastai commented:
Bit confused that tests are failing on a part of Dependabot not touched by this PR (it tries to use `pandoc` which is not installed).
Is CI broken on main/master or something?

jurre commented:
> Bit confused that tests are failing on a part of Dependabot not touched by this PR (it tries to use `pandoc` which is not installed). Is CI broken on main/master or something?

I don't see this failures you're referring to? I do see a few linting errors for nuget that we'll need to address, and there are a few test suites that are unfortunately sometimes flaky, if a suite other than `nuget` fails in this PR I can retry/fix it on main though.

Zastai commented:
That's strange; I don't see it now either - only a style check, on a source I don't think I changed. I'll push an update for that to see if that fixes the checks.

xlgmokha commented:
> I think @xlgmokha has some more experience on this so I would love to get his eyes on this as well.

LGTM. I'm not sure where/why pandoc is used but I'll take a peek. Otherwise...  :shipit:

xlgmokha commented:
It looks like a new version of [pandoc-ruby](https://rubygems.org/gems/pandoc-ruby/versions/2.1.5) was released recently that includes a change in behaviour that is triggering a failure in the test suite. This repo doesn't seem to include a version controlled `Gemfile.lock` so the resolved version changed and caused this failure. I'll try to get this fixed up in the main branch so that we can get these changes released.

xlgmokha commented:
I'm working on a fix in https://github.com/dependabot/dependabot-core/pull/4412 that should restore the `common` build to a healthier state. Unfortunately, that means we'll need to update this branch which is something I don't have access to do. 
@Zastai Do you mind pulling in the latest when https://github.com/dependabot/dependabot-core/pull/4412 is merged? I'll take care of the rest.

Zastai commented:
Yeah, no worries - ping me here when it lands and I'll pull in main.

Zastai commented:
@xlgmokha Main merged; hopefully that will allow all the checks to pass.

xlgmokha commented:
Thank you for your contribution @Zastai.

Zastai commented on nuget/lib/dependabot/nuget/file_updater/project_file_declaration_finder.rb:
```diff
@@ -95,6 +111,71 @@ def declaring_file
+
+        def parse_element(string, name)
+          xml = string
+          xml += "</#{name}>" unless string.end_with?("/>")
```
I do not know. It has been a long time since I created the PR, but I _think_ I copied what was already there for other processing (like `PackageReference`).
